src/lib/auth/strategies/wpcom/proxy-iframe.js

```diff
@@ -0,0 +1,136 @@
+ *
+ * Supported WPCOM oAuth flows:
+ *  - 'wpcom-proxy-iframe'   proxy iframe wpcom authentication using passed wpcom object
+ * @extends WPcomStrategy
+ */
+export default class WPcomProxyIframe extends WPcomStrategy {
+	/**
+	 * Pass the authentication type to the parent and save the auth options and access token.
+	 * @param {Object} authentication information about the authentication type and its options
+	 * @param {string} authentication.type authentication type (wpcom-oauth|wpcom-oauth-by-token)
+	 * @param {Object} authentication.options Set of options sent to auth implementations
+	 */
+	constructor( authentication ) {
+		super( authentication.type );
+		this.options = authentication.options;
+	}
+
+	/**
+	 * Method that creates a request handler which will be used when calling endpoints, it will
+	 * overwrite the placeholder `_request`.
+	 *
+	 * @returns {function} authenticated promisfied wrapper for the passed wpcom object
+	 * @private
+	 */
+	_createAuthenticatedRequest() {
+		/**
+		 * It is basically a wrapper that promisifies and adds authentication to the request
+		 * method of passed wpcom object
+		 *
+		 * @param {Object} args request parameters
+		 * @returns {Promise} authenticated promisfied wrapper for the wpcom object request
+		 */
+		return ( args ) => {
+			return new Promise( ( resolve, reject ) => {
+				if ( ! this.options.proxy.hasOwnProperty( 'request' ) ) {
+					return reject( 'Method "proxy.request" is required.' );
+				}
+
+				this.options.proxy.request( args, ( error, response ) => {
+					if ( error ) {
+						debug( `Request ${ args.method }:${ args.path } failed with : ${ error }` );
+						return reject( error );
+					}
+
+					debug( `${ args.method }:${ args.path } response ${ response }` );
+					resolve( response );
+				} );
+			} );
+		};
+	}
+
+	/**
+	 * Placeholder methods that will be overwritten by a promisified authenticated request method
+	 * after the login is successful.
+	 * @returns {Promise} rejects if the login failed and the method is not overwritten
+	 * @private
+	 */
+	_request() {
+		return Promise.reject( `Authentication ${ this.type } error: user not authenticated.` );
+	}
+
+	/**
+	 * Login method it assumes that the presence of the passed wpcom object means the customer is
+	 * authenticated. It also overwrites the `_request` method with a promisified authenticated one
+	 *
+	 * @returns {Promise} the oauth access token
+	 */
+	login() {
+		// overwrite the request method to an authenticated promisified wpcom proxy iframe request
+		this._request = this._createAuthenticatedRequest();
+
+		// consider it logged in for now
+		// TODO: check how wpcom failures work
+		return Promise.resolve();
+	}
+
+	// API endpoints
+	//////////////////////////////////////////////////////////////////////////////////////////
+
+	/**
+	 * Get current user details
+	 * @returns {Promise} current user details
+	 */
+	getUser() {
```

Comment:
Is the `/me` route different for oauth and proxy-iframe strategies? I'd be surprised if so. I guess this may be moved to the wpcom base class.

Reply:
Sadly yes, one library needs `apiNamespace` and the other one doesn't work with it. So I decided to keep them separated instead of adding weird rules.